src/librustc_parse/lexer/mod.rs

```diff
@@ -318,6 +315,36 @@ impl<'a> StringReader<'a> {
             }
         }
     }
+    fn cook_doc_comment(
+        &self,
+        content_start: BytePos,
+        content: &str,
+        comment_kind: CommentKind,
+        doc_kind: rustc_lexer::DocCommentKind,
+    ) -> TokenKind {
+        let mut idx = 0;
+        loop {
+            idx = match content[idx..].find('\r') {
+                None => break,
+                Some(it) => idx + it + 1,
+            };
+            self.err_span_(
+                content_start + BytePos(idx as u32 - 1),
+                content_start + BytePos(idx as u32),
+                match comment_kind {
+                    CommentKind::Line => "bare CR not allowed in doc-comment",
+                    CommentKind::Block => "bare CR not allowed in block doc-comment",
+                },
+            );
+        }
```

Comment:
This loop makes me sad, but zero-cost abstraction are not so zero cost here: https://godbolt.org/z/vE7xsr :(

Reply:
`char_indices`-based slow seems to at least unroll somewhat and avoids any panic-handling: https://godbolt.org/z/WP5YaP

Reply:
Hm, actually, we should just fast-path this...